File: src/praisonai/praisonai/capabilities/ocr.py
"""
OCR Capabilities Module

Provides optical character recognition functionality via LiteLLM.
"""

from dataclasses import dataclass, field
from typing import Optional, Union, List, Any, Dict


@dataclass
class OCRResult:
    """Result from OCR processing."""
    text: str
    pages: Optional[List[Dict[str, Any]]] = None
    model: Optional[str] = None
    usage: Optional[Dict[str, int]] = None
    metadata: Dict[str, Any] = field(default_factory=dict)


def ocr(
    document: Union[str, Dict[str, str]],
    model: str = "mistral/mistral-ocr-latest",
    include_image_base64: bool = False,
    pages: Optional[List[int]] = None,
    image_limit: Optional[int] = None,
    timeout: float = 600.0,
    api_key: Optional[str] = None,
    api_base: Optional[str] = None,
    metadata: Optional[Dict[str, Any]] = None,
    **kwargs
) -> OCRResult:
    """
    Extract text from documents or images using OCR.
    
    Args:
        document: Document URL, file path, or dict with type and URL
            - String: treated as URL or file path
            - Dict: {"type": "document_url", "document_url": "https://..."} or
                   {"type": "image_url", "image_url": "https://..."}
        model: Model name (e.g., "mistral/mistral-ocr-latest")
        include_image_base64: Whether to include base64 images in response
        pages: Specific pages to process (for PDFs)
        image_limit: Maximum number of images to process
        timeout: Request timeout in seconds
        api_key: Optional API key override
        api_base: Optional API base URL override
        metadata: Optional metadata for tracing
        
    Returns:
        OCRResult with extracted text
        
    Example:
        >>> result = ocr("https://example.com/document.pdf")
        >>> print(result.text)
        
        >>> result = ocr({"type": "image_url", "image_url": "https://example.com/image.png"})
    """
    import litellm
    
    # Handle document input
    doc_input = document
    if isinstance(document, str):
        # Check if it's a URL or file path
        if document.startswith(('http://', 'https://')):
            if document.lower().endswith(('.png', '.jpg', '.jpeg', '.gif', '.webp')):
                doc_input = {"type": "image_url", "image_url": document}
            else:
                doc_input = {"type": "document_url", "document_url": document}
        else:
            # Local file - read and encode as base64
            import base64
            with open(document, 'rb') as f:
                content = base64.b64encode(f.read()).decode('utf-8')
            if document.lower().endswith(('.png', '.jpg', '.jpeg', '.gif', '.webp')):
                doc_input = {"type": "image_url", "image_url": f"data:image/png;base64,{content}"}
            else:
                doc_input = {"type": "document_url", "document_url": f"data:application/pdf;base64,{content}"}
    
    call_kwargs = {
        'model': model,
        'document': doc_input,
        'timeout': timeout,
    }
    
    if include_image_base64:
        call_kwargs['include_image_base64'] = include_image_base64
    if pages:
        call_kwargs['pages'] = pages
    if image_limit:
        call_kwargs['image_limit'] = image_limit
    if api_key:
        call_kwargs['api_key'] = api_key
    if api_base:
        call_kwargs['api_base'] = api_base
    
    call_kwargs.update(kwargs)
    
    if metadata:
        call_kwargs['metadata'] = metadata
    
    response = litellm.ocr(**call_kwargs)
    
    # Extract text from pages
    text_parts = []
    pages_data = []
    if hasattr(response, 'pages'):
        for page in response.pages:
            page_text = getattr(page, 'markdown', '') or getattr(page, 'text', '')
            text_parts.append(page_text)
            pages_data.append({
                'index': getattr(page, 'index', 0),
                'markdown': getattr(page, 'markdown', ''),
                'images': getattr(page, 'images', []),
            })
    
    usage = None
    if hasattr(response, 'usage_info'):
        usage = {
            'pages_processed': getattr(response.usage_info, 'pages_processed', 0),
            'doc_size_bytes': getattr(response.usage_info, 'doc_size_bytes', 0),
        }
    
    return OCRResult(
        text='\n\n'.join(text_parts),
        pages=pages_data if pages_data else None,
        model=model,
        usage=usage,
        metadata=metadata or {},
    )


async def aocr(
    document: Union[str, Dict[str, str]],
    model: str = "mistral/mistral-ocr-latest",
    include_image_base64: bool = False,
    pages: Optional[List[int]] = None,
    image_limit: Optional[int] = None,
    timeout: float = 600.0,
    api_key: Optional[str] = None,
    api_base: Optional[str] = None,
    metadata: Optional[Dict[str, Any]] = None,
    **kwargs
) -> OCRResult:
    """
    Async: Extract text from documents or images using OCR.
    
    See ocr() for full documentation.
    """
    import litellm
    
    doc_input = document
    if isinstance(document, str):
        if document.startswith(('http://', 'https://')):
            if document.lower().endswith(('.png', '.jpg', '.jpeg', '.gif', '.webp')):
                doc_input = {"type": "image_url", "image_url": document}
            else:
                doc_input = {"type": "document_url", "document_url": document}
        else:
            import base64
            with open(document, 'rb') as f:
                content = base64.b64encode(f.read()).decode('utf-8')
            if document.lower().endswith(('.png', '.jpg', '.jpeg', '.gif', '.webp')):
                doc_input = {"type": "image_url", "image_url": f"data:image/png;base64,{content}"}
            else:
                doc_input = {"type": "document_url", "document_url": f"data:application/pdf;base64,{content}"}
    
    call_kwargs = {
        'model': model,
        'document': doc_input,
        'timeout': timeout,
    }
    
    if include_image_base64:
        call_kwargs['include_image_base64'] = include_image_base64
    if pages:
        call_kwargs['pages'] = pages
    if image_limit:
        call_kwargs['image_limit'] = image_limit
    if api_key:
        call_kwargs['api_key'] = api_key
    if api_base:
        call_kwargs['api_base'] = api_base
    
    call_kwargs.update(kwargs)
    
    if metadata:
        call_kwargs['metadata'] = metadata
    
    response = await litellm.aocr(**call_kwargs)
    
    text_parts = []
    pages_data = []
    if hasattr(response, 'pages'):
        for page in response.pages:
            page_text = getattr(page, 'markdown', '') or getattr(page, 'text', '')
            text_parts.append(page_text)
            pages_data.append({
                'index': getattr(page, 'index', 0),
                'markdown': getattr(page, 'markdown', ''),
                'images': getattr(page, 'images', []),
            })
    
    usage = None
    if hasattr(response, 'usage_info'):
        usage = {
            'pages_processed': getattr(response.usage_info, 'pages_processed', 0),
            'doc_size_bytes': getattr(response.usage_info, 'doc_size_bytes', 0),
        }
    
    return OCRResult(
        text='\n\n'.join(text_parts),
        pages=pages_data if pages_data else None,
        model=model,
        usage=usage,
        metadata=metadata or {},
    )

File: src/praisonai-ts/src/agent/ocr.ts
/**
 * OCRAgent - Optical Character Recognition agent
 * 
 * Python parity with praisonaiagents/agent/ocr_agent.py
 * Extracts text from documents and images using AI models.
 */

// ============================================================================
// Configuration Types
// ============================================================================

/**
 * Configuration for OCR settings.
 * Python parity with OCRConfig dataclass.
 */
export interface OCRConfig {
  /** Include base64 images in response */
  includeImageBase64?: boolean;
  /** Specific pages to extract (for PDFs) */
  pages?: number[];
  /** Maximum images per page */
  imageLimit?: number;
  /** Timeout in seconds */
  timeout?: number;
  /** Custom API endpoint URL */
  apiBase?: string;
  /** API key for the provider */
  apiKey?: string;
}

/**
 * Page result from OCR extraction.
 */
export interface OCRPage {
  /** Page index (0-based) */
  index: number;
  /** Extracted text as markdown */
  markdown: string;
  /** Base64 images if requested */
  images?: string[];
}

/**
 * Result of OCR extraction.
 */
export interface OCRResult {
  /** Combined extracted text */
  text: string;
  /** Per-page results */
  pages: OCRPage[];
  /** Additional metadata */
  metadata?: Record<string, any>;
}

/**
 * Request handed to an {@link OCRExtractor}.
 */
export interface OCRExtractRequest {
  /** URL or path the caller asked to read. */
  source: string;
  /** Provider document reference derived from `source`. */
  document: { type: string; [key: string]: string };
  /** Model to use for this extraction. */
  model: string;
  /** Resolved OCR options for this extraction. */
  options: Required<OCRConfig>;
  /** Custom API endpoint URL, if configured. */
  baseUrl?: string;
  /** API key, if configured. */
  apiKey?: string;
}

/**
 * A pluggable OCR backend. Supplying one is the only way to make
 * `OCRAgent.extract()` return text.
 */
export type OCRExtractor = (request: OCRExtractRequest) => Promise<OCRResult>;

/**
 * Configuration for creating an OCRAgent.
 */
export interface OCRAgentConfig {
  /**
   * OCR backend that performs the extraction.
   *
   * Without one, `extract()` throws. It used to return
   * `"[OCR extraction from <source> - requires API integration]"` and log
   * `✓ OCR complete`, so a caller could not tell a fabricated document
   * from a real one.
   */
  extractor?: OCRExtractor;
  /** Agent name */
  name?: string;
  /** Optional instructions */
  instructions?: string;
  /** LLM model (e.g., "mistral/mistral-ocr-latest") */
  llm?: string;
  /** Alias for llm parameter */
  model?: string;
  /** Custom API endpoint URL */
  baseUrl?: string;
  /** API key for the provider */
  apiKey?: string;
  /** OCR configuration */
  ocr?: boolean | OCRConfig;
  /** Verbosity level for output */
  verbose?: boolean | number;
}

// ============================================================================
// Default Configuration
// ============================================================================

const DEFAULT_OCR_CONFIG: Required<OCRConfig> = {
  includeImageBase64: false,
  pages: [],
  imageLimit: 0,
  timeout: 600,
  apiBase: '',
  apiKey: '',
};

// ============================================================================
// OCRAgent Class
// ============================================================================

/**
 * A specialized agent for OCR (Optical Character Recognition).
 * 
 * Extracts text from documents (PDFs) and images using AI models.
 * 
 * Supported Providers:
 * - Mistral: `mistral/mistral-ocr-latest`
 * 
 * @example
 * ```typescript
 * import { OCRAgent } from 'praisonai';
 * 
 * const agent = new OCRAgent({ llm: 'mistral/mistral-ocr-latest' });
 * 
 * // Extract from PDF URL
 * const result = await agent.extract('https://example.com/document.pdf');
 * console.log(result.text);
 * 
 * // Extract from image URL
 * const result2 = await agent.extract('https://example.com/image.png');
 * for (const page of result2.pages) {
 *   console.log(page.markdown);
 * }
 * ```
 */
export class OCRAgent {
  static readonly DEFAULT_MODEL = 'mistral/mistral-ocr-latest';

  readonly name: string;
  private readonly instructions?: string;
  private readonly llm: string;
  private readonly baseUrl?: string;
  private readonly apiKey?: string;
  private readonly ocrConfig: Required<OCRConfig>;
  private readonly verbose: boolean | number;
  private readonly extractor?: OCRExtractor;

  constructor(config: OCRAgentConfig) {
    // Handle model alias
    const llm = config.llm ?? config.model;
    
    this.name = config.name || 'OCRAgent';
    this.instructions = config.instructions;
    this.llm = llm || process.env.PRAISONAI_OCR_MODEL || OCRAgent.DEFAULT_MODEL;
    this.baseUrl = config.baseUrl;
    this.apiKey = config.apiKey;
    this.verbose = config.verbose ?? true;
    this.extractor = config.extractor;

    // Resolve OCR configuration
    this.ocrConfig = this.resolveOCRConfig(config.ocr);
  }

  private resolveOCRConfig(ocr?: boolean | OCRConfig): Required<OCRConfig> {
    if (ocr === undefined || ocr === true || ocr === false) {
      return { ...DEFAULT_OCR_CONFIG };
    }
    return { ...DEFAULT_OCR_CONFIG, ...ocr };
  }

  private buildDocument(source: string): { type: string; [key: string]: string } {
    const sourceLower = source.toLowerCase();
    const imageExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.webp'];
    
    if (imageExtensions.some(ext => sourceLower.includes(ext))) {
      return { type: 'image_url', image_url: source };
    }
    return { type: 'document_url', document_url: source };
  }

  private log(message: string): void {
    if (this.verbose) {
      console.log(message);
    }
  }

  /**
   * Extract text from a document or image.
   * 
   * @param source - URL or path to document/image
   * @param options - Override options for this extraction
   * @returns OCRResult with pages, markdown content, and metadata
   */
  async extract(
    source: string,
    options?: {
      includeImageBase64?: boolean;
      pages?: number[];
      imageLimit?: number;
      model?: string;
    }
  ): Promise<OCRResult> {
    const model = options?.model || this.llm;
    
    this.log(`Extracting text with ${model}...`);

    // Build document reference
    const document = this.buildDocument(source);

    // No backend, no result. This deliberately throws instead of returning a
    // shaped placeholder: the previous code produced
    // `"[OCR extraction from <source> - requires API integration]"`, logged
    // `✓ OCR complete`, and handed downstream code something that looked
    // exactly like a successful extraction.
    if (!this.extractor) {
      throw new Error(
        `OCRAgent has no extractor configured, so '${source}' was not read. ` +
          "Pass one explicitly, e.g. `new OCRAgent({ extractor: myMistralOcrCall })`. " +
          `Intended model: ${model}.`
      );
    }

    const result = await this.extractor({
      source,
      document,
      model,
      options: {
        ...this.ocrConfig,
        includeImageBase64: options?.includeImageBase64 ?? this.ocrConfig.includeImageBase64,
        pages: options?.pages ?? this.ocrConfig.pages,
        imageLimit: options?.imageLimit ?? this.ocrConfig.imageLimit,
      },
      baseUrl: this.baseUrl,
      apiKey: this.apiKey,
    });

    this.log('✓ OCR complete');
    return result;
  }

  /**
   * Async version of extract() - same implementation since extract is already async.
   */
  async aextract(
    source: string,
    options?: {
      includeImageBase64?: boolean;
      pages?: number[];
      imageLimit?: number;
      model?: string;
    }
  ): Promise<OCRResult> {
    return this.extract(source, options);
  }

  /**
   * Quick OCR - extract and return markdown text.
   * 
   * @param source - URL or path to document/image
   * @returns Extracted text as markdown string
   */
  async read(source: string): Promise<string> {
    const result = await this.extract(source);
    
    // Combine all pages into markdown
    if (result.pages && result.pages.length > 0) {
      return result.pages
        .map(page => page.markdown)
        .filter(Boolean)
        .join('\n\n');
    }
    return result.text;
  }

  /**
   * Async version of read().
   */
  async aread(source: string): Promise<string> {
    return this.read(source);
  }
}

// ============================================================================
// Factory Function
// ============================================================================

/**
 * Create an OCRAgent instance.
 * 
 * @param config - OCRAgent configuration
 * @returns OCRAgent instance
 */
export function createOCRAgent(config: OCRAgentConfig): OCRAgent {
  return new OCRAgent(config);
}
